### PR TITLE
feat(vm): support wasm api async

### PR DESCRIPTION
**Pull request by huangzhiran**

### CODE REVIEW

File: pkg/modules/vm/api/server.go [hunk 1]
```diff
@@ -1,24 +1,90 @@
...
+
+	router.GET("/system/hello", hello)
+
+	return router
+}
+
+func NewServer(ctx context.Context) *Server {
+	router := newRouter()
+
+	redisConf := types.MustRedisEndpointFromContext(ctx)
```
Comment (CoderZhi) [inline pkg/modules/vm/api/server.go:64]: redis endpoint could be parameter of `NewServer`

File: pkg/modules/vm/api/server.go [hunk 2]
```diff
@@ -1,24 +1,90 @@
...
+	asyncCli := asynq.NewClient(redisCli)
+	asyncSrv := asynq.NewServer(redisCli, asynq.Config{})
+	mux := asynq.NewServeMux()
+
+	mux.Handle(async.TaskNameApiCall, async.NewApiCallProcessor(ctx, router, asyncCli))
+	mux.Handle(async.TaskNameApiResult, async.NewApiResultProcessor(ctx))
+
+	l := types.MustLoggerFromContext(ctx)
+	_, l = l.Start(ctx, "vm.api.NewServer")
+	defer l.End()
```
Comment (CoderZhi) [inline pkg/modules/vm/api/server.go:78]: this is the only place `l` is needed, which you may move out of function `NewServer`

File: pkg/modules/vm/api/server.go [hunk 3]
```diff
@@ -1,24 +1,90 @@
...
 
-	srv.HandleFunc("/system/hello", hello)
+	if err := asyncSrv.Start(mux); err != nil {
+		l.Fatal(err)
```
Comment (CoderZhi) [inline pkg/modules/vm/api/server.go:81]: return error

File: pkg/modules/vm/api/server.go [hunk 4]
```diff
@@ -1,24 +1,90 @@
...
+	cli    *asynq.Client
+	srv    *asynq.Server
 }
 
-func (s *Server) Serve(req *http.Request) *http.Response {
-	resp := httptest.NewRecorder()
-	s.srv.ServeHTTP(resp, req)
-	return resp.Result()
+func (s *Server) Call(projectName string, data []byte) *http.Response {
+	l := types.MustLoggerFromContext(s.ctx)
```
Comment (CoderZhi) [inline pkg/modules/vm/api/server.go:24]: context should be the parameter of function `Call`, and `Logger` is not a `must`

File: pkg/modules/vm/api/async/processor.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+	req, err := http.ReadRequest(bufio.NewReader(bytes.NewBuffer(payload.Data)))
+	if err != nil {
+		return fmt.Errorf("http.ReadRequest failed: %v: %w", err, asynq.SkipRetry)
+	}
+
+	resp := httptest.NewRecorder()
+	p.router.ServeHTTP(resp, req)
+
+	l := types.MustLoggerFromContext(p.sysCtx)
```
Comment (CoderZhi) [inline pkg/modules/vm/api/async/processor.go:49]: use `ctx` instead of `sysCtx`

File: pkg/modules/vm/api/async/processor.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+type ApiCallProcessor struct {
+	sysCtx context.Context
+	router *gin.Engine
+	cli    *asynq.Client
+}
+
+func NewApiCallProcessor(ctx context.Context, router *gin.Engine, cli *asynq.Client) *ApiCallProcessor {
+	return &ApiCallProcessor{
+		sysCtx: ctx,
```
Comment (CoderZhi) [inline pkg/modules/vm/api/async/processor.go:29]: delete

File: pkg/modules/vm/api/async/processor.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	return nil
+}
+
+type ApiResultProcessor struct {
+	sysCtx context.Context
+}
+
+func NewApiResultProcessor(ctx context.Context) *ApiResultProcessor {
+	return &ApiResultProcessor{
+		sysCtx: ctx,
```
Comment (CoderZhi) [inline pkg/modules/vm/api/async/processor.go:85]: delete

File: pkg/modules/vm/api/async/processor.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	}
+}
+
+func (p *ApiResultProcessor) ProcessTask(ctx context.Context, t *asynq.Task) error {
+	payload := apiResultPayload{}
+	if err := json.Unmarshal(t.Payload(), &payload); err != nil {
+		return fmt.Errorf("json.Unmarshal failed: %v: %w", err, asynq.SkipRetry)
+	}
+
+	sysCtx := types.WithProject(p.sysCtx, &models.Project{
```
Comment (CoderZhi) [inline pkg/modules/vm/api/async/processor.go:95]: use `ctx` instead

File: pkg/modules/vm/api/server.go [hunk 9]
```diff
@@ -1,24 +1,90 @@
...
+	defer l.End()
 
-	srv.HandleFunc("/system/hello", hello)
+	if err := asyncSrv.Start(mux); err != nil {
+		l.Fatal(err)
+	}
 
-	return &Server{srv}
+	return &Server{
+		ctx:    ctx,
```
Comment (CoderZhi) [inline pkg/modules/vm/api/server.go:85]: delete

File: pkg/modules/vm/api/server.go (hunk 3 above)
Comment (huangzhiran) [reply pkg/modules/vm/api/server.go:81]: the asyncSrv is strong dependence，maybe would be better fatal if it starts failed

Comment (CoderZhi) [reply pkg/modules/vm/api/server.go:79]: even through, the error handling should be here: https://github.com/machinefi/w3bstream/pull/697/files#diff-d54c5acd4c6f127d45bd07e933e7a628ffb46793e21126d6e4843d9c4fafacccR162

File: pkg/modules/vm/api/async/processor.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	ctx = types.WithMgrDBExecutor(ctx, p.mgrDB)
+	ctx = types.WithLogger(ctx, p.l)
+	ctx = types.WithProject(ctx, &models.Project{
+		ProjectName: models.ProjectName{Name: payload.ProjectName}},
+	)
+
+	_, l := p.l.Start(ctx, "vm.api.ProcessTaskApiResult")
+	defer l.End()
```
Comment (CoderZhi) [inline pkg/modules/vm/api/async/processor.go:105]: `l` must be from `ctx`

File: pkg/modules/vm/api/async/processor.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+	if err != nil {
+		return fmt.Errorf("http.ReadRequest failed: %v: %w", err, asynq.SkipRetry)
+	}
+
+	resp := httptest.NewRecorder()
+	p.router.ServeHTTP(resp, req)
+
+	_, l := p.l.Start(context.Background(), "vm.api.ProcessTaskApiCall")
+	defer l.End()
+	l = l.WithValues("ProjectName", payload.ProjectName)
```
Comment (CoderZhi) [inline pkg/modules/vm/api/async/processor.go:53]: `l` must from `ctx`

File: pkg/modules/vm/wasmtime/instance_exports.go [hunk 12]
```diff
@@ -151,13 +149,9 @@ func (ef *ExportFuncs) ApiCall(kAddr, kSize, vmAddrPtr, vmSizePtr int32) int32 {
...
 
-	req, err := http.ReadRequest(bufio.NewReader(bytes.NewBuffer(buf)))
-	if err != nil {
-		ef.logAndPersistToDB(conflog.ErrorLevel, efSrc, err.Error())
-		return int32(wasm.ResultStatusCode_ParamIllegal)
-	}
+	prj := types.MustProjectFromContext(ef.ctx)
 
-	resp := ef.srv.Serve(req)
+	resp := ef.srv.Call(prj.Name, buf)
```
Comment (CoderZhi) [inline pkg/modules/vm/wasmtime/instance_exports.go:154]: use ef.ctx

File: pkg/modules/vm/api/server.go [hunk 13]
```diff
@@ -1,24 +1,87 @@
...
+	srv *asynq.Server
 }
 
-func (s *Server) Serve(req *http.Request) *http.Response {
-	resp := httptest.NewRecorder()
-	s.srv.ServeHTTP(resp, req)
-	return resp.Result()
+func (s *Server) Call(projectName string, data []byte) *http.Response {
+	_, l := s.l.Start(context.Background(), "vm.api.Call")
+	defer l.End()
```
Comment (CoderZhi) [inline pkg/modules/vm/api/server.go:26]: add `ctx` to `Call` and read `l` from `ctx`

File: pkg/modules/vm/api/server.go [hunk 14]
```diff
@@ -1,24 +1,87 @@
...
+
+	"github.com/machinefi/w3bstream/pkg/depends/conf/log"
+	"github.com/machinefi/w3bstream/pkg/depends/conf/redis"
+	"github.com/machinefi/w3bstream/pkg/depends/kit/sqlx"
+	"github.com/machinefi/w3bstream/pkg/modules/vm/api/async"
 )
 
 type Server struct {
-	srv *http.ServeMux
+	l   log.Logger
```
Comment (CoderZhi) [inline pkg/modules/vm/api/server.go:19]: delete

File: pkg/modules/vm/api/server.go [hunk 15]
```diff
@@ -1,24 +1,87 @@
 package api
```
Comment (CoderZhi) [inline pkg/modules/vm/api/server.go:1]: rename to wasmapi

File: pkg/modules/vm/wasmapi/async/processor.go [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+		return fmt.Errorf("encode http response failed: %v: %w", err, asynq.SkipRetry)
+	}
+
+	eventType := req.Header.Get("eventType")
+	if eventType == "" {
+		l.Error(errors.New("miss eventType"))
+		return fmt.Errorf("miss eventType, projectName %v: %w", payload.ProjectName, asynq.SkipRetry)
+	}
+
+	task, err := newApiResultTask(payload.ProjectName, eventType, wbuf.Bytes())
```
Comment (Liuhaai) [inline pkg/modules/vm/wasmapi/async/processor.go:65]: the result of some tasks isn't necessary, for example, logging

File: pkg/modules/vm/wasmapi/async/processor.go [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+	"github.com/pkg/errors"
+
+	"github.com/machinefi/w3bstream/pkg/depends/kit/sqlx"
+	"github.com/machinefi/w3bstream/pkg/models"
+	"github.com/machinefi/w3bstream/pkg/modules/event"
+	"github.com/machinefi/w3bstream/pkg/types"
+)
+
+type ApiCallProcessor struct {
+	router *gin.Engine
```
Comment (Liuhaai) [inline pkg/modules/vm/wasmapi/async/processor.go:23]: what's the purpose of `gin` engine here

File: pkg/modules/vm/wasmapi/async/processor.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+	ctx = types.WithMgrDBExecutor(ctx, p.mgrDB)
+	ctx = types.WithProject(ctx, &models.Project{
+		ProjectName: models.ProjectName{Name: payload.ProjectName}},
+	)
+
+	l := types.MustLoggerFromContext(ctx)
+	_, l = l.Start(ctx, "wasmapi.ProcessTaskApiResult")
+	defer l.End()
+
+	if _, err := event.HandleEvent(ctx, payload.EventType, payload.Data); err != nil {
```
Comment (Liuhaai) [inline pkg/modules/vm/wasmapi/async/processor.go:103]: how to prevent infinite recursion?

File: pkg/modules/vm/wasmapi/async/processor.go (hunk 16 above)
Comment (huangzhiran) [reply pkg/modules/vm/wasmapi/async/processor.go:65]: the result is http.Response. return the result back to caller would be better. let caller decide

File: pkg/modules/vm/wasmapi/async/processor.go (hunk 17 above)
Comment (huangzhiran) [reply pkg/modules/vm/wasmapi/async/processor.go:23]: gin is more friendly user then go http. and can use gin middleware for param validate

File: pkg/modules/vm/wasmapi/async/processor.go (hunk 18 above)
Comment (huangzhiran) [reply pkg/modules/vm/wasmapi/async/processor.go:103]: good point. Maybe we cannot prevent cyclic call in the server. we can add warn at the user guide

**Review by hunshenshi — APPROVED** [review]

**Pull request merged**